In [2]:
true_labels = [0, 0, 1, 1, 2, 2]
predicted_labels = [0, 1, 1, 1, 2, 0]
num_classes = 3

class_names = {0: "yếu", 1: "trung bình", 2: "tốt"}

In [3]:
 #bai1
correct = 0
total = len(true_labels)
for i in range(total):
    if true_labels[i] == predicted_labels[i]:
        correct += 1

accuracy = correct / total

print(f"Số lượng mẫu dự đoán đúng: {correct}/{total}")
print(f"Mô hình đạt chỉ số Accuracy cơ bản = {accuracy:.4f} (tương đương {accuracy * 100:.1f}%)")

Số lượng mẫu dự đoán đúng: 4/6
Mô hình đạt chỉ số Accuracy cơ bản = 0.6667 (tương đương 66.7%)


1. Accuracy KHÔNG phản ánh đầy đủ chất lượng của một mô hình phân loại
2. Chỉ số này sẽ gây hiểu lầm nghiêm trọng trong các trường hợp sau:
- Dữ liệu bị mất cân bằng nghiêm trọng (Imbalanced Data): Khi một lớp chiếm đa số
Khi hậu quả của việc bỏ sót (Sai lầm loại 1 và loại 2) có mức độ nghiêm trọng khác nhau
  (Ví dụ: Dự đoán bệnh nhân mắc ung thư hoặc phát hiện giao dịch gian lận).

Tổng số mẫu: 100 (Lớp 0: 95 mẫu, Lớp 1: 5 mẫu)
Nếu mô hình luôn luôn đoán kết quả là lớp 0, Accuracy đạt được là: 0.9500 (95.0%)

Dù mô hình đạt độ chính xác cực kỳ cao (95%), mô hình này hoàn toàn VÔ DỤNG.
Nó hoàn toàn thất bại trong việc tìm ra bất kỳ mẫu nào thuộc lớp 1 (chỉ có 5 mẫu).
Điều này chứng minh rằng chỉ dựa vào một mình chỉ số Accuracy là chưa đủ


phan B

1. Confusion Matrix (Ma trận nhầm lẫn) là một bảng thống kê giúp đánh giá
   hiệu năng của mô hình phân loại bằng cách so sánh kết quả dự đoán với thực tế.
2. Quy ước cấu trúc ma trận trong bài học này
   - Mỗi HÀNG : Đại diện cho các nhãn THỰC TẾ (True Labels).
   - Mỗi CỘT : Đại diện cho các nhãn DỰ ĐOÁN (Predicted Labels).


In [6]:
#Bài 5]
num_classes = 3

confusion_matrix = []
for i in range(num_classes):
    row = []
    for j in range(num_classes):
        row.append(0)
    confusion_matrix.append(row)


print("Ma trận confusion_matrix rỗng vừa được khởi tạo:")
for row in confusion_matrix:
    print(row)

Ma trận confusion_matrix rỗng vừa được khởi tạo:
[0, 0, 0]
[0, 0, 0]
[0, 0, 0]


In [7]:
#bai6
for i in range(len(true_labels)):
    true = true_labels[i]
    pred = predicted_labels[i]
    confusion_matrix[true][pred] += 1

In [8]:
#bai7
class_mapping = {0: "Hàng lớp 0 (Thực tế YẾU)      ",
                 1: "Hàng lớp 1 (Thực tế TRUNG BÌNH)",
                 2: "Hàng lớp 2 (Thực tế TỐT)      "}

for i in range(num_classes):
    print(f"{class_mapping[i]}: {confusion_matrix[i]}")

Hàng lớp 0 (Thực tế YẾU)      : [1, 1, 0]
Hàng lớp 1 (Thực tế TRUNG BÌNH): [0, 2, 0]
Hàng lớp 2 (Thực tế TỐT)      : [1, 0, 1]


In [11]:
#bài 8
total_elements_sum = 0

for r in range(num_classes):
    for c in range(num_classes):
        total_elements_sum += confusion_matrix[r][c]

actual_total_samples = len(true_labels)

print(f"Tổng tất cả các phần tử trong Confusion Matrix = {total_elements_sum}")
print(f"Tổng số lượng mẫu dữ liệu đầu vào thực tế     = {actual_total_samples}")

if total_elements_sum == actual_total_samples:
    print("hop le")
else:
    print("loi")

Tổng tất cả các phần tử trong Confusion Matrix = 6
Tổng số lượng mẫu dữ liệu đầu vào thực tế     = 6
hop le


bai9
Precision của một lớp đo lường tỷ lệ số mẫu dự đoán ĐÚNG trên tổng số mẫu được DỰ ĐOÁN thuộc lớp đó.
 Trong số các mẫu mô hình gắn nhãn là lớp A

Recall của một lớp đo lường tỷ lệ số mẫu dự đoán ĐÚNG trên tổng số mẫu THỰC TẾ thuộc lớp đó.

- True Positive (TP) : Thực tế là Lớp 1 và mô hình đoán ĐÚNG là Lớp 1.
- False Positive (FP): Thực tế LÀ LỚP KHÁC nhưng mô hình đoán SAI thành Lớp 1 (Bị nhận vơ / Dương tính giả).
- False Negative (FN): Thực tế LÀ LỚP 1 nhưng mô hình đoán SAI thành lớp khác (Bị bỏ sót / Âm tính giả).


In [15]:
#bai12

target_class = 1

tp_c1 = confusion_matrix[target_class][target_class]

fp_c1 = 0
for r in range(num_classes):
    fp_c1 += confusion_matrix[r][target_class]
fp_c1 = fp_c1 - tp_c1


fn_c1 = 0
for c in range(num_classes):
    fn_c1 += confusion_matrix[target_class][c]
fn_c1 = fn_c1 - tp_c1

print(f"True Positive (TP)  = {tp_c1}")
print(f"False Positive (FP) = {fp_c1}")
print(f"False Negative (FN) = {fn_c1}")

True Positive (TP)  = 2
False Positive (FP) = 1
False Negative (FN) = 0


In [17]:
#bai13
def precision_for_class(matrix, class_index):
    tp = matrix[class_index][class_index]

    total_predicted = 0
    for r in range(len(matrix)):
        total_predicted += matrix[r][class_index]

    if total_predicted == 0:
        return 0.0

    return tp / total_predicted

# Chạy thử nghiệm cho lớp 1
p_1 = precision_for_class(confusion_matrix, 1)
print(f"Precision của lớp 1 = {p_1:.4f}")

Precision của lớp 1 = 0.6667


In [18]:
#bai14
def recall_for_class(matrix, class_index):
    tp = matrix[class_index][class_index]


    total_actual = 0
    for c in range(len(matrix[class_index])):
        total_actual += matrix[class_index][c]


    if total_actual == 0:
        return 0.0

    return tp / total_actual

# Chạy thử nghiệm cho lớp 1
r_1 = recall_for_class(confusion_matrix, 1)
print(f"Recall của lớp 1 = {r_1:.4f}")

Recall của lớp 1 = 1.0000


In [19]:
#bai15
class_precisions = []
class_recalls = []

for c in range(num_classes):
    p = precision_for_class(confusion_matrix, c)
    r = recall_for_class(confusion_matrix, c)
    class_precisions.append(p)
    class_recalls.append(r)

    print(f"Lớp {c} ({class_names[c].upper()}): Precision = {p:.4f} | Recall = {r:.4f}")

Lớp 0 (YẾU): Precision = 0.5000 | Recall = 0.5000
Lớp 1 (TRUNG BÌNH): Precision = 0.6667 | Recall = 1.0000
Lớp 2 (TỐT): Precision = 1.0000 | Recall = 0.5000


 Lớp mô hình học TỐT NHẤT: Lớp 1 (Trung bình) vì cả Precision (0.6667) và Recall (1.0000) đều ở mức cao, đặc biệt Recall = 1.0 nghĩa là không bỏ sót bất kỳ mẫu trung bình nào.
- Lớp mô hình học KÉM NHẤT : Lớp 0 (Yếu) và Lớp 2 (Tốt).
  + Lớp 0 có Recall thấp (0.5000), nghĩa là bị bỏ sót mất một nửa số lượng sinh viên yếu.
  + Lớp 2 có Precision thấp (0.5000), nghĩa là khi đoán một sinh viên đạt loại 'Tốt' thì độ tin cậy chỉ có 50%, bị nhận nhầm nhiều.

phan D

1. Macro Average là phương pháp tính chỉ số trung bình bằng cách lấy tổng chỉ số
   (Precision hoặc Recall) của từng lớp rồi chia đều cho số lượng lớp.
2. Đặc điểm cốt lõi: Nó đánh giá mô hình dựa trên góc nhìn 'đẳng cấp giữa các lớp'.
   Mỗi lớp đều đóng góp trọng số bằng nhau vào kết quả cuối cùng, bất kể lớp đó
   có chứa 10 mẫu hay 10,000 mẫu dữ liệu.

In [23]:
#bai18

total_precision = 0
for p in class_precisions:
    total_precision += p

macro_precision = total_precision / num_classes
print(f"Danh sách Precision từng lớp: {[round(x, 4) for x in class_precisions]}")
print(f"=> Chỉ số Macro Precision tổng quan = {macro_precision:.4f}")

Danh sách Precision từng lớp: [0.5, 0.6667, 1.0]
=> Chỉ số Macro Precision tổng quan = 0.7222


In [25]:
#bai19

total_recall = 0
for r in class_recalls:
    total_recall += r

macro_recall = total_recall / num_classes
print(f"Danh sách Recall từng lớp: {[round(x, 4) for x in class_recalls]}")
print(f"Chỉ số Macro Recall tổng quan = {macro_recall:.4f}")

Danh sách Recall từng lớp: [0.5, 1.0, 0.5]
Chỉ số Macro Recall tổng quan = 0.6667


In [27]:
#bai20
total_tp = 0
total_fp = 0
total_fn = 0
for c in range(num_classes):
    tp = confusion_matrix[c][c]

    fp = 0
    for r in range(num_classes):
        fp += confusion_matrix[r][c]
    fp -= tp

    fn = 0
    for col in range(num_classes):
        fn += confusion_matrix[c][col]
    fn -= tp

    total_tp += tp
    total_fp += fp
    total_fn += fn

micro_precision = total_tp / (total_tp + total_fp)
micro_recall = total_tp / (total_tp + total_fn)

print(f"Tổng số lượng các điểm: TP_total = {total_tp} | FP_total = {total_fp} | FN_total = {total_fn}")
print(f" Chỉ số Micro Precision = {micro_precision:.4f}")
print(f" Chỉ số Micro Recall    = {micro_recall:.4f}")

Tổng số lượng các điểm: TP_total = 4 | FP_total = 2 | FN_total = 2
 Chỉ số Micro Precision = 0.6667
 Chỉ số Micro Recall    = 0.6667


1. Khi tập dữ liệu có sự MẤT CÂN BẰNG nghiêm trọng giữa các lớp (Imbalanced Data),
     ĐỒNG THỜI mô hình hoạt động rất tốt trên lớp này nhưng lại rất tệ trên lớp khác.
2. Mối liên hệ bản chất thực tế:
Nếu mô hình học cực tốt lớp đa số (nhiều dữ liệu) nhưng học cực tệ lớp thiểu số (ít dữ liệu):
Chỉ số Micro sẽ rất CAO (vì nó bị áp đảo bởi số lượng lớn các mẫu đúng của lớp đa số).
Chỉ số Macro sẽ rất THẤP (vì nó phạt nặng mô hình do hiệu năng ở lớp thiểu số kém).


phan e


In [29]:
#bai22



true_labels_real = [0]*10 + [1]*20 + [2]*10
predicted_labels_real = [
    0, 0, 0, 0, 0, 1, 1, 0, 2, 0,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 2, 2, 1, 1, 1, 1,
    2, 2, 2, 2, 2, 1, 1, 0, 2, 2   ]


num_classes_real = 3
print(f"Số lượng mẫu thực tế trong tập test: {len(true_labels_real)}")
print(f"Số lượng mẫu dự đoán tương ứng      : {len(predicted_labels_real)}")

Số lượng mẫu thực tế trong tập test: 40
Số lượng mẫu dự đoán tương ứng      : 40


In [31]:
#bai23
confusion_matrix_real = []
for i in range(num_classes_real):
    row = []
    for j in range(num_classes_real):
        row.append(0)
    confusion_matrix_real.append(row)

for i in range(len(true_labels_real)):
    t = true_labels_real[i]
    p = predicted_labels_real[i]
    confusion_matrix_real[t][p] += 1


class_labels = {0: "Thực tế YẾU      ", 1: "Thực tế TRUNG BÌNH", 2: "Thực tế TỐT      "}
print("Cột dự đoán tương ứng: [Lớp 0, Lớp 1, Lớp 2]")
for i in range(num_classes_real):
    print(f"{class_labels[i]}: {confusion_matrix_real[i]}")

Cột dự đoán tương ứng: [Lớp 0, Lớp 1, Lớp 2]
Thực tế YẾU      : [7, 2, 1]
Thực tế TRUNG BÌNH: [2, 16, 2]
Thực tế TỐT      : [1, 2, 7]


In [34]:
#bai24

# Tính Accuracy
correct_real = sum(1 for i in range(len(true_labels_real)) if true_labels_real[i] == predicted_labels_real[i])
accuracy_real = correct_real / len(true_labels_real)

print(f" Accuracy tổng quan của mô hình: {accuracy_real:.4f} ({accuracy_real*100:.1f}%)\n")

real_precisions = []

for c in range(num_classes_real):
    p_c = precision_for_class(confusion_matrix_real, c)
    r_c = recall_for_class(confusion_matrix_real, c)
    real_precisions.append(p_c)
    real_recalls.append(r_c)
    print(f"   - Lớp {c}: Precision = {p_c:.4f} | Recall = {r_c:.4f}")

# Tính Macro Average
macro_p_real = sum(real_precisions) / num_classes_real
macro_r_real = sum(real_recalls) / num_classes_real

print(f"   - Macro Precision = {macro_p_real:.4f}")
print(f"   - Macro Recall    = {macro_r_real:.4f}")

 Accuracy tổng quan của mô hình: 0.7500 (75.0%)

   - Lớp 0: Precision = 0.7000 | Recall = 0.7000
   - Lớp 1: Precision = 0.8000 | Recall = 0.8000
   - Lớp 2: Precision = 0.7000 | Recall = 0.7000
   - Macro Precision = 0.7333
   - Macro Recall    = 2.2000


1 Accuracy đạt 77.5%. Con số này dễ gây đánh lừa vì nó nhìn có vẻ cao,
     nhưng thực chất là do mô hình đang ăn theo lớp 'Trung bình' chiếm số lượng áp đảo (20/40 mẫu).
2. Phân tích Metric quan trọng nhất:
  RECALL của Lớp 0 (Yếu) là quan trọng nhất.
  Nếu Recall lớp 0 thấp (như kết quả trên là 0.7000), nghĩa là có 30% bị mất
  ko được hỗ trợ

Phần F

bai26
 trong bài toán y tế phát hiện bệnh hiểm nghèo như ung thư, chỉ số Recall (Độ bao phủ) đóng vai trò quan trọng hơn nhiều so với Precision.
vì lỗi bỏ sót (False Negative) . Nếu Recall thấp, mô hình sẽ bỏ sót những người thực sự có bệnh và kết luận họ khỏe mạnh. Người bệnh không được chữa trị kịp thời, dẫn đến nguy hiểm trực tiếp tới tính mạng.
Ngược lại, nếu Precision thấp, mô hình chỉ gặp lỗi  (Dương tính giả). Nghĩa là một người khỏe mạnh bị đoán nhầm là có bệnh.  

bai27
1. vì chỉ số này rất dễ gây đánh lừa khi dữ liệu bị mất cân bằng nghiêm trọng. Nếu một lớp chiếm đại đa số dữ liệu, mô hình chỉ cần luôn luôn đoán mò vào lớp đó là đã đạt Accuracy rất cao, nhưng thực chất mô hình hoàn toàn vô dụng vì bỏ sót toàn bộ lớp còn lại.
2. Precision tập trung vào chất lượng dự đoán
Recall tập trung vào số lượng tìm được
3. Ma trận này cho ta trực quan về các lỗi 'nhầm lẫn' của mô hình. Nhìn vào bảng, ta không chỉ biết mô hình đoán sai bao nhiêu mẫu, mà còn biết cụ thể lớp nào đang bị hệ thống nhìn nhầm sang lớp nào.
4.  Chúng ta nên dùng Macro Average khi tập dữ liệu bị mất cân bằng nhưng bối cảnh bài toán đòi hỏi tất cả các lớp phải có vai trò và trọng số đánh giá ngang nhau, không cho phép lớp có nhiều dữ liệu áp đảo lớp thiểu số.
5. Thay vì chỉ nhìn vào một con số Accuracy hời hợt,  biết cách xây dựng toàn bộ hệ thống đánh giá đa chiều (Precision, Recall, Macro/Micro) để hiểu rõ bản chất toán học phía sau, từ đó nhận diện được chính xác điểm mạnh, điểm yếu thực sự của mô hình học sâu."